core/controllers/diagnostic_test_player.py

```diff
@@ -77,30 +104,52 @@ def get(self, topic_id: str) -> None:
 
         diagnostic_test_skill_ids = topic.skill_ids_for_diagnostic_test
 
-        # From each skill, two questions were fetched. The first question
-        # (main question) will be presented to the learner initially and if the
-        # learner attempted incorrectly, then another question from the same
-        # skill (backup question) will be presented otherwise not. Hence the
-        # number of questions is twice the length of skill IDs.
-        question_count = len(diagnostic_test_skill_ids) * 2
-
+        skill_id_to_all_questions_fetchable_at_one_time = (
+            collections.defaultdict(list))
+        question_ids_for_fetched_skills = []
+
+        for skill_id in diagnostic_test_skill_ids:
+            questions = question_services.get_questions_by_skill_ids(
+                feconf.MAX_QUESTIONS_FETCHABLE_AT_ONE_TIME,
+                [skill_id],
+                require_medium_difficulty=True
+            )
+            skill_id_to_all_questions_fetchable_at_one_time[skill_id] = (
+                questions)
+
+            question_ids_for_fetched_skills.extend(
+                [question.id for question in questions])
+
+        # A dict with skill ID as key and a list with a maximum of two
+        # questions as value. Among the two questions, one question should be
+        # considered as the main and the other should be considered as
+        # the backup.
         skill_id_to_questions_map: Dict[
             str, List[question_domain.Question]] = (
                 collections.defaultdict(list))
 
-        questions = question_services.get_questions_by_skill_ids(
-            question_count,
-            diagnostic_test_skill_ids,
-            require_medium_difficulty=True
-        )
-
-        for question in questions:
-            linked_skill_ids = question.linked_skill_ids
-            diagnostic_test_linked_skill_ids = list(
-                set(linked_skill_ids) & set(diagnostic_test_skill_ids))
-
-            for skill_id in diagnostic_test_linked_skill_ids:
-                skill_id_to_questions_map[skill_id].append(question)
+        for skill_id, questions in (
+            skill_id_to_all_questions_fetchable_at_one_time.items()):
+            question_ids_for_other_diagnostic_test_skills = list(
+                set(question_ids_for_fetched_skills) -
+                set(question.id for question in questions)
+            )
+            # A list of question IDs, where each ID belongs to the questions
+            # that are already presented in the test or that belong to other
+            # skills. These IDs should not be considered in order to avoid the
+            # repetition of questions in the diagnostic test.
+            question_ids_to_exclude = (
+                question_ids_for_other_diagnostic_test_skills +
```

Comment:
This isn't what I meant, actually. It's not an issue to use a question in the test that happens to test some other skill as well.

What I'm thinking is, just construct skill_id_to_questions_dict directly. Something like this (which assumes you already have the skill IDs):

```
    skill_id_to_fetched_question_ids = get 2 questions for each skill
    question_ids_used_already = set()

    skill_ids_with_missing_questions = []
    for skill_id, fetched_question_ids in skill_id_to_fetched_question_ids:
        selected_question_ids = []
        for question_id in fetched_question_ids:
             if question_id in questions_used_already or question_id in excluded_question_ids:
                 continue
             question_ids_used_already.add(question_id)
             selected_question_ids.append(question_id)

             if len(selected_question_ids) >= 2:
                 break

        skill_id_to_questions_dict[skill_id] = selected_question_ids
        if len(selected_question_ids) < 2:
            skill_ids_with_missing_questions.append(skill_id)

    for skill_id in skill_ids_with_missing_questions:
        fetch (say) 20 questions for that skill
        populate missing parts in skill_id_to_questions_dict[skill_id] with those questions until there's 2; ignore any questions that have already been used (i.e. question_ids_used_already). Make sure to still update question_ids_used_already when adding a new question.
```

This is just a rough general idea, I haven't tested the code.

How does this sound?

Reply:
Thank you @seanlip for the pseudocode, I think we don't need to fetch questions twice if we fetch say (20) questions in the first place.

```python

excluded_question_ids = The question IDs that were already encountered in the test (received from frontend).
diagnostic_test_skill_ids = Diagnostic test skill IDs for a topic.

skill_id_to_questions = collections.defaultdict(list)
for skill_id in diagnostic_test_skill_ids:
    questions = question_services.get_questions_by_skill_ids(
        feconf.MAX_QUESTIONS_FETCHABLE_AT_ONE_TIME,
        [skill_id],
        require_medium_difficulty=True
    )

    for question in questions:
        if question.id in excluded_question_ids:
            continue

        if len(skill_id_to_questions[skill_id]) < 2:
            skill_id_to_questions[skill_id].append(question)
            excluded_question_ids.append(question.id)
        else:
            break

```

Please take a look at this pseudocode.
Thank you

Reply:
What is MAX_QUESTIONS_FETCHABLE_AT_ONE_TIME and what determines its value? This isn't defined anywhere so I don't know how to interpret it.

If it's a pretty large limit then I think your suggestion is indeed better and cleaner, thanks!

Also, what will you do if you don't have enough questions for a particular skill ID? Should be clear about that too -- theoretically it shouldn't happen but we should handle that gracefully somehow in the frontend.

Reply:
Thank you @seanlip, the value for MAX_QUESTIONS_FETCHABLE_AT_ONE_TIME is 20, and it is currently existing in the codebase. 
Basically, it limits the number of questions that can be fetched at a time.

For the case when a skill does not have two unique questions, then the backend logic (not present in the above pseudocode) will skip that skill for testing.

_[I am keeping myself assigned for implementing the above pseudocode.]_

Reply:
Updated

Reply:
@Nik-09 If the backend logic skips that skill for testing then what will the frontend do in such a case?

Reply:
In the frontend, for every topic, a [DiagnosticTestCurrentTopicStatusModel](https://github.com/oppia/oppia/blob/develop/core/templates/pages/diagnostic-test-player-page/diagnostic-test-current-topic-status.model.ts#L31) is created. Now if any skill is skipped from the backend, then model will not contain that skill in [_skillIdToQuestionsDict](https://github.com/oppia/oppia/blob/develop/core/templates/pages/diagnostic-test-player-page/diagnostic-test-current-topic-status.model.ts#L39) property. Hence the question corresponding to that skill will not be tested.

The case when a topic has single diagnostic test skill and due to lack of questions, that skill is skipped from the backend, then the empty dict will be passed from the backend. In this scenario, the frontend should not create the empty model so it will select the next topic ([selectNextTopicIdToTest](https://github.com/oppia/oppia/blob/develop/core/templates/pages/diagnostic-test-player-page/diagnostic-test-topic-tracker.model.ts#L187l)).
_[The creation of model is handled in the diagnostic test engine, so I will update the case there.]_

Please provide your thoughts on this.

Thank you.

Reply:
Adding to this: While assigning a skill to a topic for the diagnostic test, we check whether the skill has 2 questions or not. Do you think that we need to increase this value? I need some guidance on this because in the earlier comments you mentioned the chances for repetition is less. So I need to reiterate this part, that we have done earlier on the topic-editor side.

Reply:
I don't really understand this part:

> The case when a topic has single diagnostic test skill and due to lack of questions, that skill is skipped from the backend, then the empty dict will be passed from the backend. In this scenario, the frontend should not create the empty model so it will select the next topic ([selectNextTopicIdToTest](https://github.com/oppia/oppia/blob/develop/core/templates/pages/diagnostic-test-player-page/diagnostic-test-topic-tracker.model.ts#L187l)).

What actually happens conceptually (from the overall user-behaviour point of view)? Is that topic just not tested at all, is it made "not eligible", etc.?

-----

> Adding to this: While assigning a skill to a topic for the diagnostic test, we check whether the skill has 2 questions or not. Do you think that we need to increase this value? I need some guidance on this because in the earlier comments you mentioned the chances for repetition is less. So I need to reiterate this part, that we have done earlier on the topic-editor side.

That's a good question. Yeah, let's make it 3. That feels more robust.

I'm also fine with 5 if you think that's better. More than that and it might become too much of a burden/blocker, though.

Reply:
**For the first**: Those topics will be skipped from testing, which does not have at least a single skill with 2 unique questions.
**For the second**: Yes, 3 sounds reasonable. I have updated it.